design-proposals/helm-deployment/helmDeploymentRequirements.md

```diff
@@ -0,0 +1,999 @@
+    - The design for the same will be proposed in a different design document.
+- Secret configuration management of backend storage
+    - Currently the username and password are sensitive information for storage backend and are visible in plain text.
+    - They will be made available to the controller and node plugins using the Secrets resource of Kubernetes.
+    - Use Helm for secret create and update
+- Based on the Kubernetes version, configure standard sidecars for controller + node plugin. 
+    - Make sidecar configurable based on Kubernetes version
+    - Allow customer to configure sidecars
+- Maintain Helm and Kubernetes version compatibility
+- Add example for minimum viable Helm deployment
+
+
+#### Functional Requirements
+- Ensure standard eSDK deployment with Helm
+- Version compatibility between eSDK, Kubernetes and Helm.
+#### Non Functional Requirements
+
+##### Performance Requirements
+- Helm deployment should not take considerable amount of time.
+
+##### Security Requirements
+- Keep username and password secure
+
+##### Other Non Functional Requirements (Scalability, HA etc…)
+- Image size should be be large.
+- Multi-architecture support
+- Ensure deployment of single esdk instance in a Kubernetes cluster
+
+## Architecture Analysis
+
+### System Architecture
+
+![Helm deployment of eSDK architecture](helm_esdk_deployment_view.jpg)
+
+- Current deployment approach
+    - User has to understand and make multiple configuration changes
+    - There are no default attributes / values defined
+    - The complexity of software lifecycle management wrt upgrade, rollback etc is cumbersome.
+    
+- This begs to adoption of a more robust, scalable and standard deployment approach.
+    - Helm is the industry accepted standard for package deployment and management for Kubernetes.
+
+- Helm will act as a client and will abstract:
+    - software lifecycle management
+    - configuration management
+
+- Helm will initiate the installation, upgrade and rollback by converting the templates into the respective resource instances and deploy them over K8S cluster.
+
+- Helm will support the uninstallation of the release.
+
+### Architecture Tenets
+N/A
+
+
+### High Level Module Architecture
+N/A  
+
+## Detailed Design
+
+
+### Helm charts location
+
+- The Helm charts will be maintained in the GitHub repo for eSDK.
+- The charts and templates with the values will be maintained in the same repository at the same level as `src` as follows:  
+
+![Helm eSDK Code placement](helm_esdk_repository_placement.png)
+
+- This folder contains samples with some examples of basic deployments
+
+### Helm repository
+<!-- - If time permits and required -->
+
+### Configuration Management
+Helm uses a packaging format called charts. A chart is a collection of files that describe a related set of Kubernetes 
+resources. Charts are created as files laid out in a particular directory tree. They can be packaged into versioned 
+archives to be deployed.
```

Comment:
cosmetic comment: "They can be packaged into versioned archives to be deployed." --> "The same directory tree can be archived and used for deployment"

Reply:
Done